### PR TITLE
[MONIT-27738] Add rate limiting by bytes to logging for the proxy

### PR DESCRIPTION
**Pull request by xuranchen**

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java [hunk 1]
```diff
@@ -114,7 +116,7 @@ public void run() {
     isSending = true;
     try {
       List<T> current = createBatch();
-      int currentBatchSize = current.size();
+      int currentBatchSize = rateLimitOnBytes ? current.size() : getBatchSizeInBytes(current);
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java:119]: self comment from Sean: TODO: flip this check

File: proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java [hunk 2]
```diff
@@ -185,8 +187,18 @@ protected List<T> createBatch() {
     List<T> current;
     int blockSize;
     synchronized (mutex) {
+      int rateLimit = (int)rateLimiter.getRate();
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java:190]: Add comment that you considered storing the bytes and opted for the method of least code change to start and should we hit cpu issues, this is a potential area of re-visit

**Review by amitw-vmware — APPROVED** [review]
LGTM

File: proxy/src/main/java/com/wavefront/agent/ProxyConfig.java [hunk 3]
```diff
@@ -224,7 +225,7 @@ public class ProxyConfig extends Configuration {
 
   @Parameter(names = {"--pushRateLimitLogs"}, description = "Limit the outgoing logs " +
       "rate at the proxy. Default: do not throttle.")
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfig.java:227]: nit: put a note here to say it's in bytes

File: proxy/src/main/java/com/wavefront/agent/ProxyConfig.java [hunk 4]
```diff
@@ -194,9 +197,9 @@ public class ProxyConfig extends Configuration {
       "in a single flush. Default: 50")
   int pushFlushMaxEvents = DEFAULT_BATCH_SIZE_EVENTS;
 
-  @Parameter(names = {"--pushFlushMaxLogs"}, description = "Maximum allowed logs " +
-      "in a single flush. Default: 50")
-  int pushFlushMaxLogs = DEFAULT_BATCH_SIZE_LOGS;
+  @Parameter(names = {"--pushFlushMaxLogs"}, description = "Maximum size of a log payload " +
+      "in a single flush between 1mb and 5mb. Default: 4mb")
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfig.java:201]: nit: update to include maybe in parenthesis that (40,000)

File: proxy/src/main/java/com/wavefront/agent/data/EntityPropertiesFactoryImpl.java [hunk 5]
```diff
@@ -384,13 +384,13 @@ protected String getRateLimiterName() {
...
     @Override
-    public int getItemsPerBatchOriginal() {
+    public int getDataPerBatchOriginal() {
       return wrapped.getPushFlushMaxLogs();
     }
 
     @Override
     public int getMemoryBufferLimit() {
-      return 16 * wrapped.getPushMemoryBufferLimitLogs();
+      return wrapped.getPushMemoryBufferLimitLogs();
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/data/EntityPropertiesFactoryImpl.java:393]: thanks for fixing!

File: proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java [hunk 6]
```diff
@@ -258,7 +258,25 @@ public void drainBuffersToQueue(@Nullable QueueingReason reason) {
...
-        (isSending ? properties.getItemsPerBatch() / 2 : 0));
+    return getDataSize(datum) + (isBuffering.get() ? properties.getMemoryBufferLimit() :
+        (isSending ? properties.getDataPerBatch() / 2 : 0));
+  }
+
+  /**
+   * @param datum     list from which to calculate the sub-list
+   * @param ratelimit the rate limit
+   * @param batchSize the size of the batch
+   * @return index i such that datum[0:i) falls within the rate limit
```
Comment (amitw-vmware) [inline proxy/src/main/java/com/wavefront/agent/handlers/AbstractSenderTask.java:269]: nit: returns size of sublist

File: proxy/src/main/java/com/wavefront/agent/ProxyConfig.java [hunk 7]
```diff
@@ -194,9 +197,9 @@ public class ProxyConfig extends Configuration {
       "in a single flush. Default: 50")
   int pushFlushMaxEvents = DEFAULT_BATCH_SIZE_EVENTS;
 
-  @Parameter(names = {"--pushFlushMaxLogs"}, description = "Maximum allowed logs " +
-      "in a single flush. Default: 50")
-  int pushFlushMaxLogs = DEFAULT_BATCH_SIZE_LOGS;
+  @Parameter(names = {"--pushFlushMaxLogs"}, description = "Maximum size of a log payload " +
+      "in a single flush between 1mb (1048576) and 5mb (5242880). Default: 4mb (4194304)")
```
Comment (joannak-vmware) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfig.java:201]: nit: could you add to comment that these numbers in parenthesis are bytes

**Review by joannak-vmware — APPROVED** [review]
LGTM

**Review by sbhakta-vmware — APPROVED** [review]
lgtm. QA certified!

**Pull request merged**